karapace/kafka_rest_apis/__init__.py

```diff
@@ -163,16 +167,52 @@ def _add_kafka_rest_routes(self) -> None:
         self.route("/topics/<topic:path>", callback=self.topic_details, method="GET", rest_request=True)
         self.route("/topics/<topic:path>", callback=self.topic_publish, method="POST", rest_request=True)
 
-    async def get_producer(self) -> AIOKafkaProducer:
-        if self.producer_queue.empty():
-            for _ in range(self.config["producer_count"]):
-                self.log.info("Creating async producers")
-                p = await self._create_async_producer()
-                await self.producer_queue.put(p)
-                self.producer_refs.append(p)
-        return await self.producer_queue.get()
+    async def maybe_initilize_producer_queue(self) -> asyncio.Queue:
+        """Lazily initialize Producers."""
+        async with self._producer_queue_lock:
+            if self._producer_queue is None:
+                producer_queue = asyncio.Queue()
+                for _ in range(self.config["producer_count"]):
+                    try:
+                        producer = await self._create_async_producer()
+                    except:
+                        # Try to behave nicely on error, this is on a best-effort basis
+                        while not producer_queue.empty():
+                            other = await producer_queue.get_nowait()
+                            await other.close()
+
+                        # Bail on error to initialize a producer, otherwise close_producers would
+                        # block forever
+                        raise
+
+                    await producer_queue.put(producer)
+
+                # Only expose the queue if all producers were sucessfully created
+                self._producer_queue = producer_queue
+
+            return self._producer_queue
+
+    async def close_producers(self) -> None:
+        """Close all producers.
+
+        Note:
+            This method blocks until every producer can be closed
+        """
+        # Avoid initializing the producers on shutdown
+        async with self._producer_queue_lock:
+            if self._producer_queue is None:
+                self._producer_queue = asyncio.Queue()
+            return
+
+        producer_queue = await self.maybe_initilize_producer_queue()
```

Comment:
Sorry, I don't get it, why do we have to call "maybe_initialize" when we close producers?

Reply:
Instead of calling the method `get_producer_queue` I named it to `maybe_initialize_producer_queue` to be explicit about the lazy behavior. The point was exactly to note that the getter has additional side effects.

With that said, I did add code to handle the special case were the queue is uninitialized. If you want I can replace this with a `producer_queue = self._producer_queue` within the critical section above.